lib/wasix/src/state/context_switching.rs

```diff
@@ -0,0 +1,476 @@
+
+        // Add the context-switching environment to the WasiEnv
+        let previous = ctx
+            .data_mut(&mut store)
+            .context_switching_environment
+            .replace(this);
+        assert!(previous.is_none()); // Should never be hit because of the check at the top
+
+        // Turn the store into an async store and run the entrypoint
+        let store_async = store.into_async();
+        let result = local_executor.run_until(entrypoint.call_async(&store_async, params));
+
+        // Process if this was terminated by a context entrypoint returning
+        let result = match &result {
+            Err(e) => match e.downcast_ref::<ContextEntrypointReturned>() {
+                Some(ContextEntrypointReturned(id)) => {
+                    // Context entrypoint returned, which is not allowed
+                    // Exit with code 129
+                    tracing::error!("The entrypoint of context {id} returned which is not allowed");
+                    Err(RuntimeError::user(
+                        WasiError::Exit(ExitCode::from(129)).into(),
+                    ))
+                }
+                _ => result,
+            },
+            _ => result,
+        };
+        tracing::trace!("Main context finished execution and returned {result:?}");
+
+        // Drop the executor to ensure all references to the StoreAsync are gone and convert back to a normal store
+        drop(local_executor);
+        let mut store = store_async.into_store().ok().unwrap();
+
+        // Remove the context-switching environment from the WasiEnv
+        let env = ctx.data_mut(&mut store);
+
+        env.context_switching_environment
+            .take()
+            .or_else(|| {
+                env.vfork
+                    .as_mut()
+                    .and_then(|vfork| vfork.env.context_switching_environment.take())
+                    .inspect(|_| {
+                        // Grace for vforks, so they don't bring everything down with them.
+                        // This is still an error.
+                        // The message below is oversimplified there is more nuance to this.
+                        tracing::error!("Exiting a vforked process in any other way than calling `_exit()` is undefined behavior but the current program just did that.");
+                    })
+            })
+            .expect("Failed to remove wasix context-switching environment from WASIX env after main context finished. This means we lost it somehow which should never happen.");
+
+        (store, result)
+    }
+
+    /// Get the ID of the currently active context
+    pub(crate) fn active_context_id(&self) -> u64 {
+        self.inner.current_context_id.load(Ordering::Relaxed)
+    }
+
+    /// Get the id of the main context (0)
+    pub(crate) fn main_context_id(&self) -> u64 {
+        MAIN_CONTEXT_ID
+    }
+
+    pub(crate) fn destroy_context(&self, target_context_id: &u64) -> bool {
```

Comment:
Add a TODO specifying that contexts aren't properly deallocated until the next switch operation and that we should fix this because it's a waste of memory